libp2p/protocols/pubsub/pubsubpeer.nim

```diff
@@ -204,20 +206,25 @@ proc connectImpl(p: PubSubPeer) {.async.} =
 proc connect*(p: PubSubPeer) =
   asyncSpawn connectImpl(p)
 
-proc sendImpl(conn: Connection, encoded: seq[byte]) {.async.} =
-  try:
-    trace "sending encoded msgs to peer", conn, encoded = shortLog(encoded)
-    await conn.writeLp(encoded)
-    trace "sent pubsub message to remote", conn
+proc sendImpl(conn: Connection, encoded: seq[byte]): Future[void] {.raises: [Defect].} =
+  trace "sending encoded msgs to peer", conn, encoded = shortLog(encoded)
 
-  except CatchableError as exc: # never cancelled
-    # Because we detach the send call from the currently executing task using
-    # asyncSpawn, no exceptions may leak out of it
-    trace "Unable to send to remote", conn, msg = exc.msg
-    # Next time sendConn is used, it will be have its close flag set and thus
-    # will be recycled
+  let fut = conn.writeLp(encoded) # Avoid copying `encoded` into future
+  proc sendWaiter(): Future[void] {.async.} =
```

Comment:
Again not sure where is the gain here, wouldn't this do exactly the same amount of copying as the previous version but in a more convoluted way? What am I missing?

Reply:
To understand this, you need to look into how the compiler transforms the code into a closure iterator - in particular, it will copy the variables that enter the closure onto a heap instance - the closure environment - which then gets passed around by reference. `encoded` is such a variable, often having significant size. 

The best place to see this is the `C` code. The comment simplifies things by calling it `into the future` - this is a mental shortcut that is mostly accurate.

before:
```
				T9_ = (tyObject_FuturecolonObjectType___GXFSekg1U8JRoedGa2vBSA*)0;				T9_ = writeLp__9c71DJYgEH8rgDWbdnwWECg(T8_, (*(*colonenvP_).colonup_).encoded2->data, ((*(*colonenvP_).colonup_).encoded2 ? (*(*colonenvP_).colonup_).encoded2->Sup.len : 0));				asgnRef((void**) (&(*colonenvP_).chronosInternalTmpFuture4), &T9_->Sup);
```

after:
```
	asgnRef((void**) (&(*colonenv_).fut1), writeLp__9c71DJYgEH8rgDWbdnwWECg(T4_, encoded->data, (encoded ? encoded->Sup.len : 0)));
```

If you want more background on how libp2p uses memory, the easiest place is the metrics - `seq[byte]`, `Future[void]` and `string` are the 3 kinds of memory that libp2p allocates a lot of and this was one example: to create the future, it first copied `encoded` into the closure environment, then passed a reference to that environment to `writeLp` - because the closure is a single allocation that has roughly the same lifetime as the future, the copy of `encoded` stayed referenced until the future was resolved. This can take a while, and the result is that we see spikes of memory usage in Nimbus.

By doing it this way, the closure environment still has a significant lifetime, but it's now much smaller - in particular, it doesn't hold on to a copy of `encoded`.

These kinds of optimizations _generally_ don't make sense unless they've been verified to cause trouble - they are rarely noticeable - but `gossipsub`, because of the fanout, is a pathological case - every message is copied to each peer and goes through 4-5 layers of abstraction each taking multiple copies of it as part of framing it - the fewer such copies, the less memory libp2p will use and keep referenced.